File: src/mem/mtree/timing_bmt.cc
#include "mem/mtree/timing_bmt.hh"

#include "base/trace.hh"
#include "debug/TimingBmt.hh"
#include "debug/TimingBmtInit.hh"
#include "mem/mtree/util.hh"

namespace gem5 {

TimingBmt::TimingBmt(unsigned int arity, uint64_t total_data) :
    arity(arity),
    hashInputSize(BLOCK_SIZE_BYTES),
    hashOutputSize(hashInputSize / arity)
{
    dataSize = 0;

    assert(arity > 0);

    // Note that this should be in sync with the Python version

    // Calculate number of MACs and nodes this requires.
    // Each cache line is turned into a MAC.
    macCount = ceilDiv(total_data, hashInputSize);
    DPRINTF(TimingBmtInit, "%s: Number of MACs: %lld\n", __func__, macCount);
    // `arity` number of MACs can fit in a single node.
    macNodes = ceilDiv(macCount, arity);
    DPRINTF(TimingBmtInit, "%s: Number of MAC nodes: %lld\n",
        __func__, macNodes);
    dataSize += macNodes;

    // Calculate number of counters and counter nodes this requires.
    // Each cache line has a counter.
    long long counter_count = ceilDiv(total_data, hashInputSize);
    DPRINTF(TimingBmtInit, "%s: Number of (minor) counters: %lld\n",
        __func__, counter_count);
    // In a split counter configuration, we consider a 64-byte / 512-bit cache
    // line size. Each "counter node" contains a 64-bit major counter and
    // enough minor counters to cover a single 4KB page.
    // Consider a 512-bit cacheline = 64-bit major counter + 448 bits of minor
    // counters.
    // If one page is 4KB, and each counter is for a 64-byte cacheline, you
    // need 64 total cachelines for a page to fit within that 448 bits. This
    // gives 7 bits for each counter.

    unsigned int minor_counter_space = (BLOCK_SIZE_BYTES * 8) -
        majorCounterBits;
    unsigned int counters_per_page = PAGE_SIZE_BYTES / hashInputSize;
    minorCounterBits = minor_counter_space / counters_per_page;
    DPRINTF(TimingBmtInit, "%s: Major counter size: %u bits\n",
        __func__, majorCounterBits);
    DPRINTF(TimingBmtInit, "%s: Minor counter size: %u bits\n",
        __func__, minorCounterBits);

    // Ensure an invalid configuration is not created.
    assert(majorCounterBits + (minorCounterBits * counters_per_page) <=
        (BLOCK_SIZE_BYTES * 8));

    counterNodes = ceilDiv(counter_count, counters_per_page);
    DPRINTF(TimingBmtInit, "%s: Number of counter nodes: %lld\n",
        __func__, counterNodes);
    dataSize += counterNodes;

    // Calculate the number of nodes used to create the tree that protects
    // counters.

    leaves = ceilDiv(counterNodes, arity);
    height = integerLog(leaves, arity) + 1;

    DPRINTF(TimingBmtInit, "%s: The height of the tree is %u\n",
        __func__, height);

    long long non_leaf_nodes = integerPower(arity, height - 1)/(arity - 1);
    dataSize += non_leaf_nodes;
    dataSize += leaves;
    treeNodes = non_leaf_nodes + leaves;

    // Make sure nothing weird happened.
    assert(treeNodes > 0);
    assert(counterNodes > 0);
    assert(macNodes > 0);
    assert(dataSize == treeNodes + counterNodes + macNodes);

    DPRINTF(TimingBmtInit, "%s: Total number of nodes is %llu.\n",
        __func__, dataSize);

    DPRINTF(TimingBmtInit, "%s: Total space protected by tree: %llu bytes. "
        "(Requested: %llu bytes.)\n",
        __func__, statDataProtected(), total_data);

    DPRINTF(TimingBmtInit,
        "%s: Total space taken by tree nodes: %llu bytes.\n",
        __func__, (treeNodes * BLOCK_SIZE_BYTES));
    DPRINTF(TimingBmtInit,
        "%s: Total space taken by counter nodes: %llu bytes.\n",
        __func__, (counterNodes * BLOCK_SIZE_BYTES));
    DPRINTF(TimingBmtInit,
        "%s: Total space taken by MAC nodes: %llu bytes.\n",
        __func__, (macNodes * BLOCK_SIZE_BYTES));
    DPRINTF(TimingBmtInit,
        "%s: Total space taken by tree: %llu bytes.\n",
        __func__, statStructureSize());
}

TimingBmt::~TimingBmt() {
}

size_t
TimingBmt::indexOfType(size_t index)
{
    auto type = getNodeType(index);

    switch (type) {
        case TreeNodeType::TreeNode:
        return index;
        break;

        case TreeNodeType::Counter:
        return index - treeNodes;
        break;

        case TreeNodeType::MAC:
        return index - treeNodes - counterNodes;
        break;

        default:
        panic("Unexpected tree node type.");
    }

    // Default return value to please the compiler.
    return 0;
}

size_t
TimingBmt::firstOfType(AbstractIntegrityTree::TreeNodeType type)
{
    switch (type) {
        case TreeNodeType::TreeNode:
        return 0;
        break;

        case TreeNodeType::Counter:
        return treeNodes;
        break;

        case TreeNodeType::MAC:
        return treeNodes + counterNodes;
        break;

        default:
        panic("Unexpected tree node type.");
    }

    // Default return value to please the compiler.
    return 0;
}

uint64_t
TimingBmt::simulatedBlockOffset(size_t index)
{
    return index * BLOCK_SIZE_BYTES;
}

size_t
TimingBmt::parentBlockIndex(size_t index)
{
    auto type = getNodeType(index);

    switch (type) {
        case TreeNodeType::TreeNode:
        // This is a more standard tree traversal using the usual methods.

        // The first node in the tree is considered the root, and has no
        // parent.
        assert(index != firstOfType(TreeNodeType::TreeNode));

        return firstOfType(TreeNodeType::TreeNode) +
            ((indexOfType(index) - 1) / arity);
        break;

        case TreeNodeType::Counter:
        // Consider the parent of a counter node to be the tree leaf node that
        // is protecting it.

        return firstOfType(TreeNodeType::TreeNode) +
            (indexOfType(index) / arity);
        break;

        case TreeNodeType::MAC:
        // For now, we will consider the "parent" of a MAC to be the counter
        // that helps generate the MAC.

        auto macs_in_node = arity;
        auto counters_in_node = PAGE_SIZE_BYTES / hashInputSize;

        // Both the list of MACs and list of counters are mapped the same way,
        // but the MACs are on a different scale from MACs. This represents the
        // number of MAC nodes that correspond to one counter node.
        assert(counters_in_node > macs_in_node);
        auto scale_factor = counters_in_node / macs_in_node;

        return firstOfType(TreeNodeType::Counter) +
            (indexOfType(index) / scale_factor);
        break;
    }

    panic("Unexpected tree node type.");
    return 0;
}

size_t
TimingBmt::addressToBlockIndex(size_t address)
{
    // Assume for now that the direct parent of a data node is the MAC.

    // Adjust the address to ensure it is aligned to the cache line size.
    address = address & ~(BLOCK_SIZE_BYTES - 1);

    size_t offset = address / (hashInputSize * arity);

    // DPRINTF(TimingTree,
        // "%s: The leaf that corresponds to this address is #%d\n",
        // __func__, offset);

    return firstOfType(TreeNodeType::MAC) + offset;
}

bool
TimingBmt::isAncestor(size_t parent, size_t child)
{
    assert(parent != child);

    while (parent < child) {
        child = parentBlockIndex(child);
        if (child < parent) {
            // Child hopped to the same level as or above the parent.
            // Not an ancestor.
            return false;
        } else if (child == parent) {
            // This is an ancestor.
            return true;
        }
    }

    return false;
}



AbstractIntegrityTree::TreeNodeType
TimingBmt::getNodeType(size_t index)
{
    assert(index < dataSize);

    // Tree nodes first, then counters, then MACs

    if (index < treeNodes) {
        return TreeNodeType::TreeNode;
    } else if (index < treeNodes + counterNodes) {
        return TreeNodeType::Counter;
    } else {
        return TreeNodeType::MAC;
    }
}



////// Statistics //////

unsigned int
TimingBmt::statTreeHeight()
{
    return height;
}

unsigned int
TimingBmt::statTreeArity()
{
    return arity;
}

unsigned int
TimingBmt::statOutputHashSize()
{
    return hashOutputSize;
}

long long
TimingBmt::statDataProtected()
{
    // NOTE Sync with Python
    return macCount * hashInputSize;
}

long long
TimingBmt::statStructureSize()
{
    // NOTE Sync with Python
    return dataSize * BLOCK_SIZE_BYTES;
}

double
TimingBmt::statDataOverheadRatio()
{
    return (double) statStructureSize() / statDataProtected();
}

} // namespace gem5

File: src/mem/mtree/TimingBmt.py
# Only used to essentially provide the Python version of a C++ function

from math import (
    ceil,
    log,
)
from typing import Tuple


def truncate_to_size_multiple(original_size: int, size: int) -> int:
    # Truncate protected local size to a multiple of a size.
    remainder = original_size % size
    if remainder != 0:
        original_size -= remainder

    return original_size


def ceil_div(numerator: int, denominator: int) -> int:
    result = int(numerator / denominator)

    if numerator % denominator != 0:
        result += 1

    return result


class TimingBmt:
    def __init__(
        self,
        arity: int,
        total_protected_data: int,
    ) -> None:
        self.arity = arity
        self.total_protected_data = total_protected_data

        self.data_size = 0

        # Note that this should be in sync with the C++ version

        self.mac_count = ceil_div(
            total_protected_data, self._get_hash_input_size()
        )
        # print(f"PYTHON: Number of MACs: {self.mac_count}")
        self.mac_nodes = ceil_div(self.mac_count, arity)
        # print(f"PYTHON: Number of MAC nodes: {self.mac_nodes}")
        self.data_size += self.mac_nodes

        self.counter_count = ceil_div(
            total_protected_data, self._get_hash_input_size()
        )
        # print(f"PYTHON: Number of (minor) counters: {self.counter_count}")

        minor_counter_space = (
            self._get_block_size_bytes() * 8
        ) - self._get_major_counter_bits()
        counters_per_page = int(
            self._get_page_size_bytes() / self._get_hash_input_size()
        )
        self.minor_counter_bits = int(minor_counter_space / counters_per_page)
        # print(
        #     f"PYTHON: Major counter size: {self._get_major_counter_bits()} bits"
        # )
        # print(f"PYTHON: Minor counter size: {self.minor_counter_bits} bits")

        assert (
            self._get_major_counter_bits()
            + (self.minor_counter_bits * counters_per_page)
        ) <= (self._get_block_size_bytes() * 8)

        self.counter_nodes = ceil_div(self.counter_count, counters_per_page)
        # print(f"PYTHON: Number of counter nodes: {self.counter_nodes}")
        self.data_size += self.counter_nodes

        self.leaves = ceil_div(self.counter_nodes, arity)
        self.height = ceil(log(self.leaves, arity)) + 1

        # print(f"PYTHON: The height of the tree is {self.height}")

        non_leaf_nodes = int(
            (self.arity ** (self.height - 1)) / (self.arity - 1)
        )
        self.data_size += non_leaf_nodes
        self.data_size += self.leaves
        self.tree_nodes = non_leaf_nodes + self.leaves

        assert self.tree_nodes > 0
        assert self.counter_nodes > 0
        assert self.mac_nodes > 0
        assert (
            self.data_size
            == self.tree_nodes + self.counter_nodes + self.mac_nodes
        )

        # print(f"PYTHON: Total number of nodes is {self.data_size}")

        # print(
        #     f"PYTHON: Total space protected by tree: {self._stat_data_protected()} (Requested: {total_protected_data} bytes.)"
        # )

        # print(
        #     f"PYTHON: Total space taken by tree nodes: {self.tree_nodes * self._get_block_size_bytes()} bytes."
        # )
        # print(
        #     f"PYTHON: Total space taken by counter nodes: {self.counter_nodes * self._get_block_size_bytes()} bytes."
        # )
        # print(
        #     f"PYTHON: Total space taken by MAC nodes: {self.mac_nodes * self._get_block_size_bytes()} bytes."
        # )
        # print(
        #     f"PYTHON: Total space taken by tree: {self._stat_structure_size()} bytes."
        # )

    def _get_block_size_bytes(self) -> int:
        return 64

    def _get_page_size_bytes(self) -> int:
        return 4096

    def _get_major_counter_bits(self) -> int:
        return 64

    def _get_hash_input_size(self) -> int:
        return self._get_block_size_bytes()

    def _get_hash_output_size(self) -> int:
        return int(self._get_hash_input_size() / self.arity)

    def _stat_data_protected(self) -> int:
        # NOTE Sync with C++

        return int(self.mac_count * self._get_hash_input_size())

    def _stat_structure_size(self) -> int:
        # NOTE Sync with C++

        return int(self.data_size * self._get_block_size_bytes())

    @staticmethod
    def determine_max_protected_size(
        min_local_size: int,
        max_local_size: int,
        total_local_size: int,
        total_remote_size: int,
        arity: int,
    ) -> Tuple[int, int]:
        """
        :param min_local_size:    Minimum amount of size that must be protected
                                  in local memory.
        :param max_local_size:    Maximum amount of size that must be protected
                                  in local memory. If this option is `0`, the
                                  theoretical maximum is `total_local_size`.
        :param total_local_size:  Total amount of memory available to work with
                                  in local memory.
        :param total_remote_size: Total amount of memory available to work
                                  with in remote (essentially static
                                  contributor to available space).
        :param arity:             The arity of the tree structure used.

        Returns the amount of data at most (in local, and remote memory) that
        can be protected based on the amount of data it would take to protect
        it, and the total amount of space available.

        Two types of memory: local and remote. ("Local" in this case means
        "the same place as integrity data." "Remote" means "the other
        place that is not the same place as integrity data.")

        This is not an actual bearing on the memory architecture used.

        NOTE: Currently assumes only one type of memory being used to store
              integrity data at a time.
        """

        # Perform a binary search to find the max amount of data to protect

        # Scale factor represents the amount of data to use in local memory.
        if min_local_size != 0:
            min_scale_factor = 1
            if max_local_size != 0:
                max_scale_factor = max_local_size / min_local_size
            else:
                max_scale_factor = total_local_size / min_local_size
        else:
            min_scale_factor = 0
            max_scale_factor = 0
        selected_scale_factor = min_scale_factor + (
            (max_scale_factor - min_scale_factor) / 2
        )
        cache_line_size = 64
        page_size = 4096

        assert total_remote_size % cache_line_size == 0

        # print(
        #     f"TimingBmt.determine_max_protected_size: min_local_size: {min_local_size}"
        # )
        # print(
        #     f"TimingBmt.determine_max_protected_size: total_local_size: {total_local_size}"
        # )
        # print(
        #     f"TimingBmt.determine_max_protected_size: total_remote_size: {total_remote_size}"
        # )
        # print(
        #     f"TimingBmt.determine_max_protected_size: min_scale_factor: {min_scale_factor}"
        # )
        # print(
        #     f"TimingBmt.determine_max_protected_size: max_scale_factor: {max_scale_factor}"
        # )

        iterations = 0
        while True:
            min_local_protected = truncate_to_size_multiple(
                int(min_local_size * min_scale_factor), cache_line_size
            )
            max_local_protected = truncate_to_size_multiple(
                int(min_local_size * max_scale_factor), cache_line_size
            )

            attempted_local_protected = int(
                min_local_size * selected_scale_factor
            )
            attempted_local_protected = truncate_to_size_multiple(
                attempted_local_protected, cache_line_size
            )

            # Calculate the amount of space we may end up protecting
            attempted_protected_size = (
                total_remote_size + attempted_local_protected
            )

            # (For integrity data)
            remaining_local_space = (
                total_local_size - attempted_local_protected
            )

            # Calculate tree size
            # print(f"TimingBmt.determine_max_protected_size: ==============")
            tree = TimingBmt(arity, attempted_protected_size)

            # print(
            #     f"TimingBmt.determine_max_protected_size: min_scale_factor: {min_scale_factor} (Local = {min_local_protected}, Remote = {total_remote_size}, Total = {min_local_protected + total_remote_size})"
            # )
            # print(
            #     f"TimingBmt.determine_max_protected_size: selected_scale_factor: {selected_scale_factor} (Local = {attempted_local_protected}, Remote = {total_remote_size}, Total = {attempted_protected_size})"
            # )
            # print(
            #     f"TimingBmt.determine_max_protected_size: --> Available space: {remaining_local_space}"
            # )
            # print(
            #     f"TimingBmt.determine_max_protected_size: --> Tree size: {tree._stat_structure_size()}"
            # )
            # print(
            #     f"TimingBmt.determine_max_protected_size: --> Difference: {remaining_local_space - tree._stat_structure_size()}"
            # )
            # print(
            #     f"TimingBmt.determine_max_protected_size: max_scale_factor: {max_scale_factor} (Local = {max_local_protected}, Remote = {total_remote_size}, Total = {max_local_protected + total_remote_size})"
            # )

            if (
                remaining_local_space - tree._stat_structure_size() >= 0
                and remaining_local_space - tree._stat_structure_size()
                < page_size
            ):
                # We have found the best arrangement of tree size with this amount of protected space. Stop.
                attempted_local_protected = truncate_to_size_multiple(
                    attempted_local_protected, page_size
                )
                print(
                    f"TimingBmt.determine_max_protected_size: Scale factor found."
                )
                return attempted_local_protected, total_remote_size
            elif tree._stat_structure_size() < remaining_local_space:
                # There is more space that could be protected, since there is leftover space on the table.
                # print(
                #     f"TimingBmt.determine_max_protected_size: Increasing scale factor."
                # )
                min_scale_factor = selected_scale_factor
            else:
                # This structure is too large for the amount of space that is leftover, reduce the protected size to make more space for integrity data
                # print(
                #     f"TimingBmt.determine_max_protected_size: Decreasing scale factor."
                # )
                max_scale_factor = selected_scale_factor

            selected_scale_factor = min_scale_factor + (
                (max_scale_factor - min_scale_factor) / 2
            )
            iterations += 1

            if iterations > 500:
                if (
                    remaining_local_space - tree._stat_structure_size()
                    >= page_size
                ):
                    print(
                        "It looks like this configuration is leaving unused space, but this is still a valid configuration."
                    )
                    return attempted_local_protected, total_remote_size

                print(
                    "Could not find a valid size arrangement for this configuration. (The integrity tree may take up too much space no matter what.)"
                )
                exit(1)
